tests/unittests/config/test_schema.py

```diff
@@ -2165,6 +2166,61 @@ class TestHandleSchemaArgs:
         "Args", "config_file schema_type docs system annotate instance_data"
     )
 
+    @pytest.mark.parametrize(
+        "failure, expected_out, expected_logs",
+        (
+            (
+                IOError("No permissions on /var/lib/cloud/instance"),
+                ["Valid schema"],
+                ["Using default instance-data/user-data paths for non-root"],
+            ),
+            (
+                DataSourceNotFoundException("No cached datasource found yet"),
+                ["Valid schema"],
+                ["datasource not detected"],
+            ),
+        ),
+    )
+    @mock.patch(M_PATH + "read_cfg_paths")
+    def test_handle_schema_unable_to_read_cfg_paths(
+        self,
+        read_cfg_paths,
+        failure,
+        expected_out,
+        expected_logs,
+        paths,
+        capsys,
+        caplog,
+        tmpdir,
+    ):
+        if isinstance(failure, IOError):
+            failure.errno = EACCES
+        read_cfg_paths.side_effect = [failure, paths]
+        user_data_fn = tmpdir.join("user-data")
+        with open(user_data_fn, "w") as f:
+            f.write(
+                dedent(
+                    """\
+                    #cloud-config
+                    packages: [sl]
+                    """
+                )
+            )
+        args = self.Args(
+            config_file=str(user_data_fn),
+            schema_type="cloud-config",
+            annotate=False,
+            docs=None,
+            system=None,
+            instance_data=None,
+        )
+        handle_schema_args("unused", args)
+        out, _err = capsys.readouterr()
```

Comment:
nit: `capsys.readouterr()` returns a namedtuple with two attributes, `out` and `err`, so this could be prettier / less boilerplate, but functionally equivalent without this line:
```python
        handle_schema_args("unused", args)
        for expected in expected_out:
            assert expected in capsys.readouterr().out
```

Reply:
yes given that we aren't providing more than one item in the expected_out list, I'll adapt this code to make the direct call to `capsys.readouterr.out` I'll also just drop the parameterization of expected_out too since we are validating the same hard-coded string for both anyway.
thx